inductor: move the CPU weight packing path after of AOTAutograd

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104614
* #104612
* #104399
* __->__ #103851

At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng

XiaobingSuper committed:
inductor: move the CPU weight packing path to after of AOTAutograd

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path to after of AOTAutograd"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path to after of AOTAutograd"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path to after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -541,10 +544,325 @@ def _register_binary_unary_fusion():
+        def forward(self, x):
+            return self.linear(self.linear(x))
+
+    the above's packed weight nodes are duplicate if two linear calls have same input size.
```
Can this duplication be avoided in the `graph.py` when the constant is added to the graph with `add_tensor_constant`?

XiaobingSuper replied:
We can't do it at ```add_tensor_constant``` which MKLDNN tensor doesn't support comparison.

jgong5 replied:
It seems the dedup logic is not necessarily specific to mkldnn weight packing patterns but can be made general, i.e., sharing the common prefix of users, e.g.,
```
w -> op1 -> op2 -> op3
  -> op1 -> op2 -> op4
```
Can be simplified to
```
w -> op1 -> op2 -> op3
                -> op4
```

XiaobingSuper commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -67,7 +69,18 @@
+}
+
 
+@config.patch({"freezing": True, "freezing_discard_parameters": True})
```
No, it just for free memory usage.

XiaobingSuper commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -271,19 +284,28 @@ def __init__(self, binary_fn, in_channels, out_channels, bias, **kwargs):
+                if len(input_shape) == 3:
+                    is_inplace = linear_binary_list[binary_fn][2]
+                    # view + linear + view(joint_graph+post_grad)
+                    match_count = match_count + 5 if is_inplace else match_count + 3
+                    match_nodes = match_nodes + 7 if is_inplace else match_nodes + 5
```
For output place op, the graph has ```view+linear+view pattern```, but for in place case, there have another two view ops. The joint_graph path will match each view op(match_counts is 2 and match_nodes is 2 for outplace, match_counts is 4 and match_nodes is 4 for inplace). And post_grad path will match the ```view+linear+view``` pattern(match_counts is 1, match_nodes is 3).

XiaobingSuper commented on torch/_inductor/graph.py:
```diff
@@ -445,7 +445,8 @@ def add_tensor_constant(self, data):
             for name, value in self.constants.items():
                 if (
-                    data.size() == value.size()
+                    not data.is_mkldnn
```
Yes, but we don't support torch.eq for MKLDNN tensor now.

XiaobingSuper commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -541,10 +544,325 @@ def _register_binary_unary_fusion():
+    for weight_idx, target in zip([1, 2], [aten.mm.default, aten.addmm.default]):
+        if (
+            node.target == target
+            and node.args[weight_idx].target == aten.permute.default
+            and node.args[weight_idx].args[0] in constant_nodes
+            and node.args[weight_idx].args[1] == [1, 0]
```
Maybe at the next step, currently, I don't consider it.

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -541,10 +544,325 @@ def _register_binary_unary_fusion():
+    return False
+
+
+def _insert_packed_convolution(gm, conv_node):
```
Changed.

XiaobingSuper commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -541,10 +544,325 @@ def _register_binary_unary_fusion():
+        gm.graph.erase_node(conv_node)
+
+
+def _insert_packed_linear(gm, linear_node):
```
Changed.

XiaobingSuper commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -541,10 +544,325 @@ def _register_binary_unary_fusion():
+    """
+    Check if the node is supported for MKLDNN convolution.
+    """
+    if node.target == aten.convolution.default and node.args[1] in constant_nodes:
```
using set now.

XiaobingSuper commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -541,10 +544,325 @@ def _register_binary_unary_fusion():
+    """
+    if (
+        not (torch.backends.mkldnn.enabled and torch.backends.mkldnn.is_available())
+        or torch.is_grad_enabled()
```
Yes, remove it.

XiaobingSuper commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -541,10 +544,325 @@ def _register_binary_unary_fusion():
             _register_binary_fusion()
+
+
+def get_preserved_arg_indices_and_constant_nodes(gm, flat_params, fw_metadata):
```
updated.

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

jgong5 reviewed (approved):
LGTM as the first step but consider to make the dedup logic more general as the next step, also need to cover more patterns like matmul without permute.

Also note that this PR requires users to explicitly set the `freezing` to `true` to enable weight prepacking, which is a change to the usage to get good performance on CPU. In the future, we'd better promote the model freezing as an official frontend API.

eellison commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -541,10 +544,298 @@ def _register_binary_unary_fusion():
+                    gm.graph.erase_node(user_node)
+
+
+def mkldnn_weight_prepack_fx(gm, flat_params, fw_metadata):
```
I need to land https://github.com/pytorch/pytorch/pull/100746, but I don't think we need this separate code path, or a lot of the changes.

As shown here:
https://github.com/pytorch/pytorch/pull/100746/files#diff-8ce7bc944676c03af78123e618ec440923b913f80d2c152db797071ff6318bc7R212-R218

we can register these pre-packing passes as a freezing pass. The `get_attr` [check](https://github.com/pytorch/pytorch/pull/100746/files#diff-4e47f07c143c95673f6877ea2d7d8778aa2222f20be3a603472a425c9ff76bbcR34-R37) is a way to check if the input is a constant, we don't need to thread constants around.

XiaobingSuper replied:
@eellison, yes， register re-packing passes as a freezing passes.

eellison commented:
Btw, also relevant: https://github.com/pytorch/pytorch/issues/103990.

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper commented:
> Btw, also relevant: #103990.

This can be next step to do that.

XiaobingSuper commented on torch/_inductor/fx_passes/freezing_patterns.py:
```diff
@@ -0,0 +1,131 @@
+"""
+@functools.lru_cache(None)
+def addmm_patterns_init():
+    if torch.cuda.is_available():
```
That code will be rebased after https://github.com/pytorch/pytorch/pull/100746 is merged.

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper commented:
@eellison, the code is rebased, please help review it.

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -541,10 +546,308 @@ def _register_binary_unary_fusion():
                     unary_attr=UnaryAttr("relu"),
                 )
 
+    def _recover_linear():
```
not reviewing this file closely since it's existing code

eellison commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -541,10 +546,308 @@ def _register_binary_unary_fusion():
+        for meta_value in [input_meta_value, weight_meta_value]:
+            if (
+                meta_value is None
+                or device_hint(meta_value) != "cpu"
```
the tensors here are fake tensors so it's more straight-forward to just call meta_value.device.type != "cpu"

XiaobingSuper replied:
changed.

eellison commented on torch/_inductor/fx_passes/freezing_patterns.py:
```diff
@@ -1,27 +1,74 @@
+
+
+def register_freezing_graph_pattern(
+    pattern, extra_check=_return_true, pass_number=0, prepend=False
```
can you remove the `prepend` call here ? it wasn't actually being used so i removed it from the arg. this will error on rebase (happened to me)

XiaobingSuper replied:
Yes, removed.

eellison reviewed (approved):
nice !! didn't look at the `mkldnn_fusion.py` too closely but looks good !!

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: move the CPU weight packing path after of AOTAutograd"


At next step:
1. support amp path for applying more fusion.
2. support dynamic shape path for applying more fusion.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge

eellison reviewed (commented):
nice !!

XiaobingSuper commented:
@pytorchbot merge
